### PR TITLE
HDDS-10568. When the ldb command is executed, it is output by line

### PR DESCRIPTION
**Pull request by jianghuazhu**
## What changes were proposed in this pull request?
When executing the ldb command, if the data is very large, a very large file will be generated, which is not friendly. This pr will add a new function that can control the maximum number of records allowed to be saved in each file.
Details:
HDDS-10568

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-10568

### CODE REVIEW

**Comment by jianghuazhu** [comment]
The new command can be combined with --pre-file-records, using:
`
./bin/ozone debug ldb --db=/xxxx/xxxx/xxxx/om.db scan --column_family=fileTable --out=/xxxx/xxxx/xxxx/om/data/tmp_fileTable --pre-file-records=1000000
`

Effective file result set:
image

**Comment by jianghuazhu** [comment]
ci :
https://github.com/jianghuazhu/ozone/actions/runs/8387158994

Can you help review this pr, @adoroszlai @xichen01 .
Thanks.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java [hunk 1]
```diff
@@ -218,10 +225,17 @@ private boolean displayTable(ManagedRocksIterator iterator,
     }
 
     // Write to file output
-    try (PrintWriter out = new PrintWriter(new BufferedWriter(
-        new PrintWriter(fileName, UTF_8.name())))) {
-      return displayTable(iterator, dbColumnFamilyDef, out, schemaV3);
+    while (iterator.get().isValid()) {
+      try (PrintWriter out = new PrintWriter(new BufferedWriter(
+          new PrintWriter(fileName + fileSuffix, UTF_8.name())))) {
```
Comment (xichen01) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:230]: If `preFileRecords` is not specified, we'd better make the filename the same as the previous filename (without `fileSuffix`)

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java [hunk 2]
```diff
@@ -288,6 +302,9 @@ private void processRecords(ManagedRocksIterator iterator,
         batch = new ArrayList<>(batchSize);
         sequenceId++;
       }
+      if ((preFileRecords > -1) && (count >= preFileRecords)) {
```
Comment (xichen01) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:305]: Seems like the `ldb` will generate unlimited empty file If the `preFileRecords` is zero.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java [hunk 3]
```diff
@@ -148,6 +150,11 @@ public class DBScanner implements Callable<Void>, SubcommandWithParent {
       defaultValue = "10")
   private int threadCount;
 
+  @CommandLine.Option(names = {"--pre-file-records"},
+      description = "The number of print records per file.",
+      defaultValue = "-1")
+  private long preFileRecords;
```
Comment (xichen01) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:156]: Nit: suggest `--max-records-per-file`

**Review by xichen01 — COMMENTED** [review]
Thanks for working on this, few comments to handle

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java (hunk 3 above)
Comment (jianghuazhu) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:156]: Thanks for your comment and review, @xichen01 .
I will update soon.

**Comment by jianghuazhu** [comment]
Can you help review this PR again, @xichen01 ?
Thanks.

**Comment by jianghuazhu** [comment]
Can you help review this pr, @kerneltime  @errose28 .
Thanks.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java [hunk 4]
```diff
@@ -218,10 +225,24 @@ private boolean displayTable(ManagedRocksIterator iterator,
...
+      while (iterator.get().isValid()) {
+        try (PrintWriter out = new PrintWriter(new BufferedWriter(
+                new PrintWriter(fileName + fileSuffix, UTF_8.name())))) {
+          if (!displayTable(iterator, dbColumnFamilyDef, out, schemaV3)) {
+            return false;
+          }
+        }
+        fileSuffix++;
+      }
+    } else {
```
Comment (xichen01) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:238]: Perhaps we can simplify this `if... else`
Like:
```
//...
String  fileNameXXX = preFileRecords > 0 ? fileName + fileSuffix++ : fileName;
//...
new PrintWriter(fileNameXXX, UTF_8.name())
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java (hunk 3 above)
Comment (adoroszlai) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:156]: Please also rename `preFileRecords` to `recordsPerFile`.

(`pre` means "before")

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java [hunk 5]
```diff
@@ -148,6 +150,11 @@ public class DBScanner implements Callable<Void>, SubcommandWithParent {
       defaultValue = "10")
   private int threadCount;
 
+  @CommandLine.Option(names = {"--max-records-per-file"},
+      description = "The number of print records per file.",
```
Comment (errose28) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:154]:
```suggestion
      description = "The number of records to print per file.",
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java [hunk 6]
```diff
@@ -288,6 +309,9 @@ private void processRecords(ManagedRocksIterator iterator,
         batch = new ArrayList<>(batchSize);
         sequenceId++;
       }
+      if ((preFileRecords > 0) && (count >= preFileRecords)) {
+        break;
+      }
```
Comment (errose28) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:314]: What's the expected behavior when this new `--max-records-per-file` flag is used without `--out`? Right now it looks like the choice that `stdout` is considered "one file" and so this flag overrides the `--length` option:
```
# The DB here has many more than 3 entries
$ ./ozone debug ldb --db=om.db scan --column_family=fileTable -l3 --max-records-per-file=2 | jq '.[].keyName' | wc -l 
2

$ ./ozone debug ldb --db=om.db scan --column_family=fileTable -l2 --max-records-per-file=3 | jq '.[].keyName' | wc -l
2
```

Maybe we should disallow `--max-records-per-file` without `--out`.

Comment (errose28) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:314]: `-l` is also broken with this new option and I got a bit of a surprise trying to test this 😄 I would have expected 5 files, here not 57 thousand.
```
$ ./ozone debug ldb --db=om.db scan --column_family=fileTable -l10 --max-records-per-file=2 --out=foo
^C
$ ls -l | grep foo | wc -l                                                                           
57343
```

**Review by errose28 — COMMENTED** [review]
Thanks for the improvement @jianghuazhu. I think the idea is solid since just using `split` on a stdout stream may produce individual files that are not valid json. Let's add some tests to `TestLDBCli` to make sure we have all the corner cases around various flag combinations working.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java (hunk 6 above)
Comment (jianghuazhu) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:314]: Thank you for your comment and review.
I will update soon.

**Comment by jianghuazhu** [comment]
Sorry, I had some other work some time ago.
I will make up for this issue in the near future.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java (hunk 6 above)
Comment (jianghuazhu) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java]: > `-l` is also broken with this new option and I got a bit of a surprise trying to test this 😄 I would have expected 5 files, here not 57 thousand.
> 
> ```
> $ ./ozone debug ldb --db=om.db scan --column_family=fileTable -l10 --max-records-per-file=2 --out=foo
> ^C
> $ ls -l | grep foo | wc -l                                                                           
> 57343
> ```

When `--out` is not set, all records are output to stdout.
When `--max-records-per-file` and `-l` exist at the same time, `--max-records-per-file` shall prevail.

**Comment by jianghuazhu** [comment]
The unit tests run fine on my local machine:
image

Can you help me look at this PR again, @errose28 @xichen01 .
Thanks.

**Comment by jianghuazhu** [comment]
https://github.com/apache/ozone/actions/runs/8849501009/job/24349282946?pr=6420
In CI, it is prompted that File-related operations cannot be performed, such as mkdirs().
Is it because it has something to do with our testing environment?
@errose28 @adoroszlai , do you know the reason?

**Comment by adoroszlai** [comment]
> In CI, it is prompted that File-related operations cannot be performed, such as mkdirs().

```
M B RV: Exceptional return value of java.io.File.mkdirs() ignored in org.apache.hadoop.ozone.debug.DBScanner.displayTable(ManagedRocksIterator, DBColumnFamilyDefinition, boolean)  At DBScanner.java:[line 232]
```

https://spotbugs.readthedocs.io/en/latest/bugDescriptions.html#rv-method-ignores-exceptional-return-value-rv-return-value-ignored-bad-practice

https://github.com/apache/ozone/blob/c7012f3bc91e92af08f2d89ee6ac1cbb103475d2/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java#L228-L233

`File.mkdirs()` returns `false` if it could not create the directories.  This return value needs to be handled (e.g. exit with error message).

```
M D NP: Possible null pointer dereference in org.apache.hadoop.ozone.debug.TestLDBCli.testScanWithRecordsPerFile() due to return value of called method  Dereferenced at TestLDBCli.java:[line 286]
```

https://spotbugs.readthedocs.io/en/latest/bugDescriptions.html#np-possible-null-pointer-dereference-due-to-return-value-of-called-method-np-null-on-some-path-from-return-value

https://github.com/apache/ozone/blob/c7012f3bc91e92af08f2d89ee6ac1cbb103475d2/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/debug/TestLDBCli.java#L285-L286

`File.listFiles` returns `null` if the path is not a directory.  Try checking if `tmpDir1` is a directory.  The same applies to `tmpDir2` below.

**Comment by jianghuazhu** [comment]
ci: https://github.com/jianghuazhu/ozone/actions/runs/8876637070
I've updated it.
Can you help me look at this PR again, @errose28 @xichen01 @adoroszlai ?
Thanks.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java [hunk 7]
```diff
@@ -305,7 +335,7 @@ private void processRecords(ManagedRocksIterator iterator,
   }
 
   private boolean withinLimit(long i) {
-    return limit == -1L || i < limit;
+    return recordsPerFile > 0 || limit == -1L || i < limit;
```
Comment (xichen01) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:338]: if the `recordsPerFile > 0` is `true`, the subsequent judgments will be short-circuited, including the `i < limit` then the limit will be invalidated. This is not a expected function.

Comment (jianghuazhu) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:338]: Thanks @xichen01  for the comment and review.
When `recordsPerFile>0`, it means that `--max-records-per-file` has taken effect, and `--limit` should be ignored at this time.

Comment (xichen01) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:338]: The `--limit` is used to limit the total count of records, the `--max-records-per-file` is used to limit the max records count of specific file.

Such as: 
` ozone debug ldb ...  --limit 10 --max-records-per-file 1 --out result.txt`

This command should generate 10 files, like result.txt0, result.txt1, ..., and each of them has 1 record.

Comment (jianghuazhu) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:338]: I'll update later.

**Comment by adoroszlai** [comment]
Continued in #7467.

**Pull request closed without merging**